### PR TITLE
Fix FETools::extrapolate  (#18389)

### LINKED ISSUE #18389: FETools::extrapolate can fail when using LinearAlgebra::distributed vectors and distributed triangulations
### Steps to reproduce the bug
This bug was first discussed on the mailing list (https://groups.google.com/g/dealii/c/bqiToP-udGY). The bug appears on deal.II >=9.6.1 when the following conditions are fulfilled:

1. The DofHandlers are based on a distributed triangulation and the simulation is done with multiple mpi ranks.
2. The output vector is a distributed vector from LinearAlgebra::distributed.
3. On any of the processors, there exists a locally owned DoF that lives on a coarse cell neighbouring a more refined ghost cell.

### Expected behavior

Calling FETools::extrapolate in the situation described above should lead to the same results (up to numerical precision) than on a serial simulation.

### Actual behavior

In the scenario described above, calling extrapolate in release mode will lead to different results than in serial simulations, and calling it in debug mode will throw an Assertion linked to the compression of the output vector https://github.com/dealii/dealii/blob/26f420caf60e81f141e6c81cae42efd74e37e4bf/include/deal.II/fe/fe_tools_extrapolate.templates.h#L1455 because of nonmatching values between a locally owned DoF and the associated ghost DoF on other processors. The cause of these nonmatching values is to be found at https://github.com/dealii/dealii/blob/26f420caf60e81f141e6c81cae42efd74e37e4bf/include/deal.II/fe/fe_tools_extrapolate.templates.h#L724-L733 which clearly expect that ghost values set by another processor on the more refined ghost neighbour should overwrite the locally owned value. This is actually what happens for Trilinos and Petsc vectors, but not for LinearAlgebra::distributed vectors due to the different compress(insert) implementation.

This bug was actually never detected in the fe_tools_extrapolate_*** tests of the testsuite because the tested triangulation was not fulfilling condition 3 above. The following patch modifies the triangulation of this test such that the bug can be easily reproduced on other developers machine without having to change the tested output (this is just an inversion symmetry operation on the refinement pattern of the triangulation):
https://github.com/warthan07/dealii/commit/1de1b85718a3a46f32af84817e6024315ad56d1d.patch

With this patch applied, this is what `ctest -j 1 -R "fe_tools_extrapolate"` returns on my pc:
```
      Start  1: all-headers/fe/fe_tools_extrapolate.templates.h.debug
 1/40 Test  #1: all-headers/fe/fe_tools_extrapolate.templates.h.debug ............   Passed    6.59 sec
      Start  2: all-headers/fe/fe_tools_extrapolate.templates.h.release
 2/40 Test  #2: all-headers/fe/fe_tools_extrapolate.templates.h.release ..........   Passed    5.91 sec
      Start  3: test_dependency/mpi.fe_tools_extrapolate_01.debug.executable
 3/40 Test  #3: test_dependency/mpi.fe_tools_extrapolate_01.debug.executable .....   Passed   14.33 sec
      Start  4: mpi/fe_tools_extrapolate_01.mpirun=1.debug
 4/40 Test  #4: mpi/fe_tools_extrapolate_01.mpirun=1.debug .......................   Passed    2.79 sec
      Start  5: test_dependency/mpi.fe_tools_extrapolate_01.release.executable
 5/40 Test  #5: test_dependency/mpi.fe_tools_extrapolate_01.release.executable ...   Passed   10.66 sec
      Start  6: mpi/fe_tools_extrapolate_01.mpirun=1.release
 6/40 Test  #6: mpi/fe_tools_extrapolate_01.mpirun=1.release .....................   Passed    1.61 sec
      Start  7: mpi/fe_tools_extrapolate_01.mpirun=3.debug
 7/40 Test  #7: mpi/fe_tools_extrapolate_01.mpirun=3.debug .......................   Passed    2.63 sec
      Start  8: mpi/fe_tools_extrapolate_01.mpirun=3.release
 8/40 Test  #8: mpi/fe_tools_extrapolate_01.mpirun=3.release .....................   Passed    1.58 sec
      Start  9: test_dependency/mpi.fe_tools_extrapolate_02.debug.executable
 9/40 Test  #9: test_dependency/mpi.fe_tools_extrapolate_02.debug.executable .....   Passed   12.69 sec
      Start 10: mpi/fe_tools_extrapolate_02.mpirun=1.debug
10/40 Test #10: mpi/fe_tools_extrapolate_02.mpirun=1.debug .......................   Passed    2.84 sec
      Start 11: test_dependency/mpi.fe_tools_extrapolate_02.release.executable
11/40 Test #11: test_dependency/mpi.fe_tools_extrapolate_02.release.executable ...   Passed   10.64 sec
      Start 12: mpi/fe_tools_extrapolate_02.mpirun=1.release
12/40 Test #12: mpi/fe_tools_extrapolate_02.mpirun=1.release .....................   Passed    1.56 sec
      Start 13: mpi/fe_tools_extrapolate_02.mpirun=3.debug
13/40 Test #13: mpi/fe_tools_extrapolate_02.mpirun=3.debug .......................   Passed    2.63 sec
      Start 14: mpi/fe_tools_extrapolate_02.mpirun=3.release
14/40 Test #14: mpi/fe_tools_extrapolate_02.mpirun=3.release .....................   Passed    1.59 sec
      Start 15: test_dependency/mpi.fe_tools_extrapolate_03.debug.executable
15/40 Test #15: test_dependency/mpi.fe_tools_extrapolate_03.debug.executable .....   Passed   13.21 sec
      Start 16: mpi/fe_tools_extrapolate_03.mpirun=1.debug
16/40 Test #16: mpi/fe_tools_extrapolate_03.mpirun=1.debug .......................   Passed    2.86 sec
      Start 17: test_dependency/mpi.fe_tools_extrapolate_03.release.executable
17/40 Test #17: test_dependency/mpi.fe_tools_extrapolate_03.release.executable ...   Passed   11.07 sec
      Start 18: mpi/fe_tools_extrapolate_03.mpirun=1.release
18/40 Test #18: mpi/fe_tools_extrapolate_03.mpirun=1.release .....................   Passed    1.61 sec
      Start 19: mpi/fe_tools_extrapolate_03.mpirun=10.debug
19/40 Test #19: mpi/fe_tools_extrapolate_03.mpirun=10.debug ......................***Failed    1.79 sec
      Start 20: mpi/fe_tools_extrapolate_03.mpirun=10.release
20/40 Test #20: mpi/fe_tools_extrapolate_03.mpirun=10.release ....................***Failed    2.06 sec
      Start 21: mpi/fe_tools_extrapolate_03.mpirun=3.debug
21/40 Test #21: mpi/fe_tools_extrapolate_03.mpirun=3.debug .......................***Failed    1.55 sec
      Start 22: mpi/fe_tools_extrapolate_03.mpirun=3.release
22/40 Test #22: mpi/fe_tools_extrapolate_03.mpirun=3.release .....................***Failed    1.79 sec
      Start 23: test_dependency/mpi.fe_tools_extrapolate_04.debug.executable
23/40 Test #23: test_dependency/mpi.fe_tools_extrapolate_04.debug.executable .....   Passed   13.76 sec
      Start 24: mpi/fe_tools_extrapolate_04.mpirun=1.debug
24/40 Test #24: mpi/fe_tools_extrapolate_04.mpirun=1.debug .......................   Passed    2.89 sec
      Start 25: test_dependency/mpi.fe_tools_extrapolate_04.release.executable
25/40 Test #25: test_dependency/mpi.fe_tools_extrapolate_04.release.executable ...   Passed   11.01 sec
      Start 26: mpi/fe_tools_extrapolate_04.mpirun=1.release
26/40 Test #26: mpi/fe_tools_extrapolate_04.mpirun=1.release .....................   Passed    1.60 sec
      Start 27: mpi/fe_tools_extrapolate_04.mpirun=3.debug
27/40 Test #27: mpi/fe_tools_extrapolate_04.mpirun=3.debug .......................   Passed    2.60 sec
      Start 28: mpi/fe_tools_extrapolate_04.mpirun=3.release
28/40 Test #28: mpi/fe_tools_extrapolate_04.mpirun=3.release .....................   Passed    1.60 sec
      Start 29: test_dependency/mpi.fe_tools_extrapolate_05.debug.executable
29/40 Test #29: test_dependency/mpi.fe_tools_extrapolate_05.debug.executable .....   Passed   14.24 sec
      Start 30: mpi/fe_tools_extrapolate_05.mpirun=1.debug
30/40 Test #30: mpi/fe_tools_extrapolate_05.mpirun=1.debug .......................   Passed    2.95 sec
      Start 31: test_dependency/mpi.fe_tools_extrapolate_05.release.executable
31/40 Test #31: test_dependency/mpi.fe_tools_extrapolate_05.release.executable ...   Passed   12.57 sec
      Start 32: mpi/fe_tools_extrapolate_05.mpirun=1.release
32/40 Test #32: mpi/fe_tools_extrapolate_05.mpirun=1.release .....................   Passed    1.64 sec
      Start 33: mpi/fe_tools_extrapolate_05.mpirun=3.debug
33/40 Test #33: mpi/fe_tools_extrapolate_05.mpirun=3.debug .......................   Passed    2.77 sec
      Start 34: mpi/fe_tools_extrapolate_05.mpirun=3.release
34/40 Test #34: mpi/fe_tools_extrapolate_05.mpirun=3.release .....................   Passed    1.64 sec
      Start 35: test_dependency/mpi.fe_tools_extrapolate_06.debug.executable
35/40 Test #35: test_dependency/mpi.fe_tools_extrapolate_06.debug.executable .....   Passed   14.20 sec
      Start 36: mpi/fe_tools_extrapolate_06.mpirun=1.debug
36/40 Test #36: mpi/fe_tools_extrapolate_06.mpirun=1.debug .......................   Passed    2.89 sec
      Start 37: test_dependency/mpi.fe_tools_extrapolate_06.release.executable
37/40 Test #37: test_dependency/mpi.fe_tools_extrapolate_06.release.executable ...   Passed   10.57 sec
      Start 38: mpi/fe_tools_extrapolate_06.mpirun=1.release
38/40 Test #38: mpi/fe_tools_extrapolate_06.mpirun=1.release .....................   Passed    1.55 sec
      Start 39: mpi/fe_tools_extrapolate_06.mpirun=10.debug
39/40 Test #39: mpi/fe_tools_extrapolate_06.mpirun=10.debug ......................***Failed    1.77 sec
      Start 40: mpi/fe_tools_extrapolate_06.mpirun=10.release
40/40 Test #40: mpi/fe_tools_extrapolate_06.mpirun=10.release ....................***Failed    1.89 sec

85% tests passed, 6 tests failed out of 40

Total Test time (real) = 216.11 sec

The following tests FAILED:
         19 - mpi/fe_tools_extrapolate_03.mpirun=10.debug (Failed)
         20 - mpi/fe_tools_extrapolate_03.mpirun=10.release (Failed)
         21 - mpi/fe_tools_extrapolate_03.mpirun=3.debug (Failed)
         22 - mpi/fe_tools_extrapolate_03.mpirun=3.release (Failed)
         39 - mpi/fe_tools_extrapolate_06.mpirun=10.debug (Failed)
         40 - mpi/fe_tools_extrapolate_06.mpirun=10.release (Failed)
```
As mentioned above, tests related to Petsc/Trilinos vectors are fine (tests 1,2,4,5), while tests related LinearAlgebra::distributed vectors fails when mpirun> 1 (tests 3,6).

### PR DESCRIPTION
**Pull request by warthan07**
This PR attempts to fix #18389. 

The key idea is first to realize that the documentation of FETools::extrapolate is inaccurate, as already recognised a few years ago but unfortunately never fixed (see https://groups.google.com/g/dealii/c/-Ne1L4fcj88/m/oikIoyi0BgAJ): on adaptively refined mesh, DoF values extrapolated on a given patch may be overridden by different values set by a children patch.

This means that, on serial triangulation and simulations, a DoF that lives both on refined and coarse patches will always have its value set by the more refined patch. This also means that we should be careful about such DoFs in distributed simulations if they live at an interface between a ghost and locally owned cell. This was precisely what https://github.com/dealii/dealii/blob/26f420caf60e81f141e6c81cae42efd74e37e4bf/include/deal.II/fe/fe_tools_extrapolate.templates.h#L724-L733 was attempting to safeguard: for such DoFs, only sets the value if the considered cell is more refined than the neighbouring ghost cells, and do a final compress(insert) operation for correctly distributing the values at the end of the calculation.

As explained in #18389, the problem is that LinearAlgebra::distributed vectors do not compress(insert) in the same way than Petsc/Trilinos vectors: whereas the latter vectors will accumulate ghost write operations in a cache and then insert these values into the locally owned DoF during the compression, the former vectors simply check that the values agree on all processors.

To fix this, I tried to mimick the compress(insert) behavior of Petsc/Trilinos vectors using a compress(max) operation and some modification on the map `dofs_on_refined_neighbors` (now renamed `dofs_at_refined_ghost_interface`) that stores the indices and cell level of DoFs that lives on a refined cell neighbouring a coarser cell, with those two neighbours owned by different processors. The basic idea is that a coarse cell should never write into such DoFs if there exists a more refined cell that will write into it in another processor. In this case, a default value of -inf is kept for this DoF, and the final compress(max) operation correctly distribute the more refined DoF value to all processors. 

I have commented my modifications so that it is not too hard to understand, but I am happy to provide further clarifications. The encouraging news is that the fe_tools_extrapolate tests (patched like in #18389 inside this PR) now all succeed whatever the choice of distributed vectors.

### CODE REVIEW

File: include/deal.II/fe/fe_tools_extrapolate.templates.h [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 #include <deal.II/base/config.h>
 
+#include "deal.II/base/types.h"
```
Comment (luca-heltai) [inline include/deal.II/fe/fe_tools_extrapolate.templates.h:21]:
```suggestion
#include <deal.II/base/types.h>
```

File: include/deal.II/fe/fe_tools_extrapolate.templates.h [hunk 2]
```diff
@@ -31,6 +32,7 @@
 
 #include <deal.II/grid/tria.h>
 
+#include "deal.II/lac/block_vector_base.h"
 #include <deal.II/lac/affine_constraints.h>
 #include <deal.II/lac/block_vector.h>
 #include <deal.II/lac/la_parallel_block_vector.h>
```
Comment (luca-heltai) [inline include/deal.II/fe/fe_tools_extrapolate.templates.h:38]:
```suggestion
#include <deal.II/lac/affine_constraints.h>
#include <deal.II/lac/block_vector.h>
#include <deal.II/lac/block_vector_base.h>
#include <deal.II/lac/la_parallel_block_vector.h>
```

File: include/deal.II/fe/fe_tools_extrapolate.templates.h [hunk 3]
```diff
@@ -1451,8 +1597,16 @@ namespace FETools
...
         }
 
-
-      u2.compress(VectorOperation::insert);
+      // Correctly synchronise values at ghost interface for all processors. For
+      // Petsc/Trilinos, a simple compress(insert) operation allows to insert
+      // values set by other processors into locally owned DoF, but the deal.II
+      // implementation of the LinearAlgebra::distributed vectors and associated
+      // compress(insert) operation does not do this. Instead, we use a
+      // compress(max) that mimick this behavior.
```
Comment (luca-heltai) [inline include/deal.II/fe/fe_tools_extrapolate.templates.h:1605]:
```suggestion
      // compress(max) that mimic this behavior.
```

**Review by luca-heltai — CHANGES_REQUESTED** [review]
Thanks! This looks already in very good shape. I'm no experti in this part of the library, so I'll let someone else comment on the actual content on the patch. Just a couple of small fixes from my side to adhere to the rest of the library.

**Comment by warthan07** [comment]
I did some additional testing of this fix for dual-weight-residual error control on my production code for liquid crystal simulation, and realised I did not anticipate all edge cases. I greatly simplified the logic of the fix in my last commit, which also fixes the edge cases encountered in the aforementioned tests. Basically my previous method was failing when a refined ghost cell was only sharing a single vertex or line with a coarser locally owned cell on a given process, and in any case, was unnecessarily complicated. 

Now the implementation is actually really close to the original one (contributed by @masterleinad back in 2016 looking at the commit history), apart from the final compress operation, which still need to be different for Petsc/Trilinos and LinearAlgebra::distributed vectors for the reason mentioned above.

File: include/deal.II/fe/fe_tools_extrapolate.templates.h [hunk 4]
```diff
@@ -697,7 +704,18 @@ namespace FETools
...
+
+    template <typename... Ts>
+    inline constexpr bool
+      is_la_vector<LinearAlgebra::distributed::Vector<Ts...>> =
+        std::true_type{};
 
+    template <typename... Ts>
+    inline constexpr bool
+      is_la_vector<LinearAlgebra::distributed::BlockVector<Ts...>> =
+        std::true_type{};
```
Comment (bangerth) [inline include/deal.II/fe/fe_tools_extrapolate.templates.h:718]: You are initializing `bool` variables here, so I think this would be clearer:
```suggestion
    template <typename T>
    inline constexpr bool is_la_vector = false;

    template <typename... Ts>
    inline constexpr bool
      is_la_vector<LinearAlgebra::distributed::Vector<Ts...>> =
        true;

    template <typename... Ts>
    inline constexpr bool
      is_la_vector<LinearAlgebra::distributed::BlockVector<Ts...>> =
        true;
```
I think it might also be nice to spell out the possible template arguments of the Vector/BlockVector class.

File: include/deal.II/fe/fe_tools_extrapolate.templates.h [hunk 5]
```diff
@@ -1339,7 +1362,14 @@ namespace FETools
 
       compute_all_non_local_data(dof2, u2_relevant);
 
-      // exclude dofs on more refined ghosted cells
+      // List all DoFs that, on this mpi rank, lives at the intersection between
```
Comment (bangerth) [inline include/deal.II/fe/fe_tools_extrapolate.templates.h:1365]:
```suggestion
      // List all DoFs that, on this mpi rank, live at the intersection between
```

File: include/deal.II/fe/fe_tools_extrapolate.templates.h [hunk 6]
```diff
@@ -1451,8 +1508,16 @@ namespace FETools
           queue.pop();
         }
 
-
-      u2.compress(VectorOperation::insert);
+      // Correctly synchronise values at ghost interface for all processors. For
```
Comment (bangerth) [inline include/deal.II/fe/fe_tools_extrapolate.templates.h:1511]: Not your fault -- we have a habit of using American English spelling:
```suggestion
      // Correctly synchronize values at ghost interface for all processors. For
```

**Review by bangerth — COMMENTED** [review]
I *think* I understand what you are doing here. Good detective work, and many thanks for the patch!

Can I ask for a couple of small changes? I have marked up 3 places, but mostly I'd like it if instead of changing the existing tests, we could duplicate them where the existing ones continue to test what they were testing, and the duplicates have the change you make to the mesh. I think that would be useful to provide as broad coverage as possible.

I think you should also put an entry into `doc/news/changes/minor` -- you should get credit for your work!

File: include/deal.II/fe/fe_tools_extrapolate.templates.h [hunk 7]
```diff
@@ -1339,7 +1362,14 @@ namespace FETools
...
 
-      // exclude dofs on more refined ghosted cells
+      // List all DoFs that, on this mpi rank, lives at the intersection between
+      // a non-artificial cell and a more refined ghost cell (the dofs being
+      // enumerated from the more refined side). For each of these DoFs, also
+      // store the most refined level among the neighboring cells. This set of
+      // DoFs naturally includes the set of DoFs shared between a refined ghost
+      // cell and a coarser locally owned cell and can be constructed without a
+      // recursion to find cells with common vertices (which are not necessarily
+      // direct neighbours in 2D and 3D).
```
Comment (masterleinad) [inline include/deal.II/fe/fe_tools_extrapolate.templates.h:1372]:
```suggestion
      // direct neighbors in 2D and 3D).
```

**Review by masterleinad — COMMENTED** [review]
Looks reasonable to me.

**Comment by warthan07** [comment]
> Can I ask for a couple of small changes? I have marked up 3 places, but mostly I'd like it if instead of changing the existing tests, we could duplicate them where the existing ones continue to test what they were testing, and the duplicates have the change you make to the mesh. I think that would be useful to provide as broad coverage as possible.

Thanks for the review! For duplicating tests, is there a particular style convention for naming the duplicate tests? fe_tools_extrapolate_alt_[01-06] or fe_tools_extrapolate_[07-12]?

**Comment by warthan07** [comment]
I have also taken the liberty of correcting the documentation of FETools::extrapolate concerning the overwriting behaviour of refined patches (as already suggested by https://groups.google.com/g/dealii/c/-Ne1L4fcj88/m/oikIoyi0BgAJ) and added my contribution statement to news/changes/minor.

**Review by luca-heltai — APPROVED** [review]

**Comment by luca-heltai** [comment]
Thanks @warthan07 !
